include/dr/detail/communicator.hpp

```diff
@@ -85,28 +93,62 @@ class communicator {
   }
 
   template <typename T>
-  void isend(const T *data, std::size_t size, std::size_t source, tag t,
+  void isend(const T *data, std::size_t count, std::size_t dest_node, tag t,
              MPI_Request *request) const {
-    MPI_Isend(data, size * sizeof(T), MPI_BYTE, source, int(t), mpi_comm_,
+    MPI_Isend(data, count * sizeof(T), MPI_BYTE, dest_node, int(t), mpi_comm_,
               request);
   }
 
   template <rng::contiguous_range R>
-  void isend(const R &data, std::size_t source, tag t,
-             MPI_Request *request) const {
-    isend(data.data(), data.size(), source, int(t), request);
+  void isend(const R &data, int dest_node, tag t, MPI_Request *request) const {
+    isend(rng::data(data), rng::size(data), dest_node, t, request);
   }
 
   template <typename T>
-  void irecv(T *data, std::size_t size, std::size_t dest, tag t,
+  void irecv(T *data, std::size_t size, std::size_t src_node, tag t,
              MPI_Request *request) const {
-    MPI_Irecv(data, size * sizeof(T), MPI_BYTE, dest, int(t), mpi_comm_,
+    MPI_Irecv(data, size * sizeof(T), MPI_BYTE, src_node, int(t), mpi_comm_,
               request);
   }
 
   template <rng::contiguous_range R>
-  void irecv(R &data, std::size_t source, tag t, MPI_Request *request) const {
-    irecv(data.data(), data.size(), source, int(t), request);
+  void irecv(R &data, std::size_t src_node, tag t, MPI_Request *request) const {
+    irecv(data.data(), data.size(), src_node, t, request);
+  }
+
+  template <rng::contiguous_range R>
+  void alltoall(R &sendr, R &recvr, std::size_t count) {
+    using T = typename R::value_type;
+    MPI_Alltoall(rng::data(sendr), count * sizeof(T), MPI_BYTE,
+                 rng::data(recvr), count * sizeof(T), MPI_BYTE, mpi_comm_);
+  }
+
+  template <typename T>
+  void alltoallv(T *sendbuf, std::vector<int> &sendcnt,
```

Comment:
I think vectors and bufs in params could be made `rng::contiguous_range`-s like they are in `alltoall`? We will have then more universal signature.

Reply:
I tried replacing bufs with ranges - but that caused compilation problems in mhp::sort, so I kept the less elegant, still more universal version. And the vectors are storing parameters (sizes, displacements) for MPI API which requires array of the type int*, just what vector<int>.data() returns - so replacing with another type them makes no sense.

Reply:
> compilation problems in mhp::sort

I'd like to see that compilation errors. Let's discuss it on person.